core/src/main/scala/kafka/server/AbstractFetcherThread.scala

```diff
@@ -205,6 +216,15 @@ abstract class AbstractFetcherThread(name: String,
     ResultWithPartitions(fetchOffsets, partitionsWithError)
   }
 
+  private def onPartitionFenced(tp: TopicPartition): Unit = {
+    Option(partitionStates.stateValue(tp)).foreach { currentFetchState =>
+      val currentLeaderEpoch = currentFetchState.currentLeaderEpoch
+      info(s"Partition $tp has an older epoch ($currentLeaderEpoch) than the current leader. Will await " +
+        s"the new LeaderAndIsr state before resuming fetching.")
+      partitionStates.remove(tp)
```

Comment:
@junrao Thanks for the suggestion. This is tricky since we would need to acquire the lock in `AbstractFetcherManager` while holding the lock in `AbstractFetcherThread`. The usual locking order is the other way around. I will explore whether it is possible to give up the lock in `AbstractFetcherThread` prior to invoking `onPartitionFenced`.